File: scripts/upload_pdf_gdrive_delta.js
import { readFileSync, writeFileSync, existsSync, readdirSync, createReadStream, mkdirSync } from "fs";
import { resolve, dirname, join } from "path";
import { fileURLToPath } from "url";
import { google } from "googleapis";
import http from "http";
import { exec } from "child_process";
import crypto from "crypto";

const __dirname = dirname(fileURLToPath(import.meta.url));

// Prioritaskan credentials dari knowledge-base, fallback ke fasih-sync-monitoring
const KB_ROOT = resolve(__dirname, "..");
const FASIH_SYNC_DIR = resolve(KB_ROOT, "..", "fasih-sync-monitoring");

const CREDENTIALS_PATH = existsSync(resolve(KB_ROOT, "credentials.json"))
  ? resolve(KB_ROOT, "credentials.json")
  : resolve(FASIH_SYNC_DIR, "credentials.json");

const USER_TOKEN_PATH = resolve(FASIH_SYNC_DIR, "token_user.json");
const GDRIVE_FOLDER_ID = "1GVLa9UVOBJOr-rb62A539HnNK7UGyrXa";
const PDF_DIR = resolve(KB_ROOT, "kegiatan/sensus-ekonomi-2026/2026/sqllab_monitoring/pdf_siap_cetak");
const LINK_MAPPING_FILE = resolve(FASIH_SYNC_DIR, "results", "pdf_gdrive_links.json");

const SCOPES = [
  "https://www.googleapis.com/auth/drive.file",
  "https://www.googleapis.com/auth/drive"
];

function getFileMD5(filePath) {
  const fileBuffer = readFileSync(filePath);
  const hashSum = crypto.createHash("md5");
  hashSum.update(fileBuffer);
  return hashSum.digest("hex");
}

async function getOAuth2Client() {
  if (!existsSync(CREDENTIALS_PATH)) {
    throw new Error(`File credentials.json tidak ditemukan di ${CREDENTIALS_PATH}`);
  }

  const keys = JSON.parse(readFileSync(CREDENTIALS_PATH, "utf-8"));
  const creds = keys.installed || keys.web;

  const oAuth2Client = new google.auth.OAuth2(
    creds.client_id,
    creds.client_secret,
    "http://localhost:8888/oauth2callback"
  );

  if (existsSync(USER_TOKEN_PATH)) {
    const token = JSON.parse(readFileSync(USER_TOKEN_PATH, "utf-8"));
    oAuth2Client.setCredentials(token);
    return oAuth2Client;
  }

  return new Promise((resolveClient, reject) => {
    const authUrl = oAuth2Client.generateAuthUrl({
      access_type: "offline",
      scope: SCOPES,
      prompt: "consent"
    });

    console.log("\n==================================================================");
    console.log("🔑 SILAKAN OTORISASI OTOMATIS BERIKUT (CUKUP 1 KALI SAJA):");
    console.log(authUrl);
    console.log("==================================================================\n");

    const server = http.createServer(async (req, res) => {
      try {
        const reqUrl = new URL(req.url, "http://localhost:8888");
        if (reqUrl.pathname === "/oauth2callback") {
          const code = reqUrl.searchParams.get("code");
          if (!code) {
            res.end("Kode otorisasi tidak ditemukan.");
            return;
          }

          const { tokens } = await oAuth2Client.getToken(code);
          oAuth2Client.setCredentials(tokens);
          writeFileSync(USER_TOKEN_PATH, JSON.stringify(tokens, null, 2));
          console.log("🟢 SUKSES! Token OAuth 2.0 User telah disimpan di token_user.json");

          res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
          res.end(`
            <div style="font-family: sans-serif; text-align: center; margin-top: 50px;">
              <h1 style="color: #2e7d32;">✅ Otorisasi Berhasil!</h1>
              <p>Token Google OAuth 2.0 milik Bg Ihza telah tersimpan.</p>
            </div>
          `);

          server.close();
          resolveClient(oAuth2Client);
        }
      } catch (err) {
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("Error: " + err.message);
        reject(err);
      }
    });

    server.listen(8888, () => {
      exec(`xdg-open "${authUrl}"`, () => {});
    });
  });
}

export async function uploadPDFsWithDeltaSync() {
  if (!existsSync(PDF_DIR)) {
    console.error(`❌ Folder PDF tidak ditemukan: ${PDF_DIR}`);
    return {};
  }

  const pdfFiles = readdirSync(PDF_DIR).filter(f => f.endsWith(".pdf"));
  if (pdfFiles.length === 0) {
    console.log("⚠️ Tidak ada berkas PDF di folder pdf_siap_cetak.");
    return {};
  }

  console.log(`📌 Memulai Delta Upload ke Google Drive (${pdfFiles.length} PDF lokal)...`);
  const auth = await getOAuth2Client();
  const drive = google.drive({ version: "v3", auth });

  // 1. Dapatkan SELURUH berkas GDrive yang ada menggunakan Pagination (pageToken)
  console.log("🔍 Fetching daftar lengkap berkas di GDrive (dengan Pagination)...");
  const existingMap = new Map();
  let pageToken = null;
  let totalRemoteFiles = 0;

  try {
    do {
      const listRes = await drive.files.list({
        q: `'${GDRIVE_FOLDER_ID}' in parents and trashed = false`,
        fields: "nextPageToken, files(id, name, webViewLink, md5Checksum)",
        pageSize: 1000,
        pageToken: pageToken,
        supportsAllDrives: true,
        includeItemsFromAllDrives: true,
      });

      for (const f of listRes.data.files || []) {
        existingMap.set(f.name, f);
        totalRemoteFiles++;
      }
      pageToken = listRes.data.nextPageToken;
    } while (pageToken);

    console.log(`  ✓ Berhasil mengambil ${totalRemoteFiles} berkas dari GDrive.`);
  } catch (e) {
    console.warn("  ⚠️ Gagal list berkas GDrive:", e.message);
  }

  const linkMapping = loadLinkMappingCache();
  let skippedCount = 0;
  let updatedCount = 0;
  let createdCount = 0;

  for (let i = 0; i < pdfFiles.length; i++) {
    const filename = pdfFiles[i];
    const codeMatch = filename.match(/^(\d{16})/);
    const subslsCode = codeMatch ? codeMatch[1] : null;
    const filePath = join(PDF_DIR, filename);

    try {
      const localMD5 = getFileMD5(filePath);

      if (existingMap.has(filename)) {
        const existingFile = existingMap.get(filename);
        const webViewLink = existingFile.webViewLink || `https://drive.google.com/file/d/${existingFile.id}/view`;

        // Pengecekan Hash MD5 (Delta Upload)
        if (existingFile.md5Checksum && existingFile.md5Checksum === localMD5) {
          if (subslsCode) linkMapping[subslsCode] = webViewLink;
          skippedCount++;
          continue;
        }

        // MD5 berbeda: Update isi file di GDrive secara In-Place (ID & Link tetap sama)
        console.log(`     🔄 [${i+1}/${pdfFiles.length}] Updating file yang berubah: ${filename}...`);
        await drive.files.update({
          fileId: existingFile.id,
          media: {
            mimeType: "application/pdf",
            body: createReadStream(filePath),
          },
          supportsAllDrives: true,
        });

        if (subslsCode) linkMapping[subslsCode] = webViewLink;
        updatedCount++;
        console.log(`       ✓ Updated in-place: ${webViewLink}`);
        continue;
      }

      // Upload file baru
      console.log(`     ↑ [${i+1}/${pdfFiles.length}] Uploading berkas baru: ${filename}...`);
      const fileMetadata = {
        name: filename,
        parents: [GDRIVE_FOLDER_ID],
      };
      const media = {
        mimeType: "application/pdf",
        body: createReadStream(filePath),
      };

      const file = await drive.files.create({
        requestBody: fileMetadata,
        media: media,
        fields: "id, name, webViewLink, md5Checksum",
        supportsAllDrives: true,
      });

      const fileId = file.data.id;
      const webViewLink = file.data.webViewLink || `https://drive.google.com/file/d/${fileId}/view`;

      try {
        await drive.permissions.create({
          fileId: fileId,
          requestBody: { role: "reader", type: "anyone" },
          supportsAllDrives: true,
        });
      } catch (pErr) {
        // ignore permission error
      }

      if (subslsCode) linkMapping[subslsCode] = webViewLink;
      existingMap.set(filename, { id: fileId, name: filename, webViewLink, md5Checksum: file.data.md5Checksum || localMD5 });
      createdCount++;
      console.log(`       ✓ Uploaded: ${webViewLink}`);
    } catch (err) {
      console.error(`       ❌ Gagal upload ${filename}: ${err.message}`);
    }
  }

  saveLinkMappingCache(linkMapping);
  console.log(`\n🟢 SUKSES DELTA UPLOAD! (Skipped/Identik: ${skippedCount}, Updated: ${updatedCount}, Baru: ${createdCount})`);
  console.log(`📌 Total ${Object.keys(linkMapping).length} PDF terhubung ke Google Drive.`);
  return linkMapping;
}

function saveLinkMappingCache(data) {
  const resultsDir = resolve(FASIH_SYNC_DIR, "results");
  if (!existsSync(resultsDir)) {
    mkdirSync(resultsDir, { recursive: true });
  }
  writeFileSync(LINK_MAPPING_FILE, JSON.stringify(data, null, 2));
}

function loadLinkMappingCache() {
  if (existsSync(LINK_MAPPING_FILE)) {
    try {
      return JSON.parse(readFileSync(LINK_MAPPING_FILE, "utf-8"));
    } catch (e) {
      return {};
    }
  }
  return {};
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  uploadPDFsWithDeltaSync().catch(console.error);
}

File: scripts/upload_pdf_gdrive_delta.py
#!/usr/bin/env python3
"""scripts/upload_pdf_gdrive_delta.py — High-Speed Thread-Safe Delta Uploader ke Google Drive."""

import os
import sys
import json
import hashlib
import requests
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed
from google.oauth2.credentials import Credentials
from google.auth.transport.requests import Request

GDRIVE_FOLDER_ID = "1GVLa9UVOBJOr-rb62A539HnNK7UGyrXa"
BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
PDF_DIR = os.path.join(BASE_DIR, "kegiatan/sensus-ekonomi-2026/2026/pdf_verifikasi_rt")
RESULTS_DIR = os.path.join(BASE_DIR, "kegiatan/sensus-ekonomi-2026/2026/outputs")
FASIH_RESULTS_DIR = os.path.abspath(os.path.join(BASE_DIR, "..", "fasih-sync-monitoring", "results"))
TOKEN_PATH = os.path.join(BASE_DIR, "token.json")

SCOPES = [
    'https://www.googleapis.com/auth/spreadsheets',
    'https://www.googleapis.com/auth/drive',
]

def get_valid_access_token():
    """Dapatkan token OAuth2 yang valid dan otomatis refresh jika expired."""
    if not os.path.exists(TOKEN_PATH):
        raise FileNotFoundError(f"token.json tidak ditemukan di {TOKEN_PATH}")
    creds = Credentials.from_authorized_user_file(TOKEN_PATH, SCOPES)
    if creds.expired and creds.refresh_token:
        creds.refresh(Request())
        with open(TOKEN_PATH, "w", encoding="utf-8") as f:
            f.write(creds.to_json())
    return creds.token

def compute_file_md5(filepath: Path) -> str:
    hasher = hashlib.md5()
    with open(filepath, 'rb') as f:
        for chunk in iter(lambda: f.read(65536), b''):
            hasher.update(chunk)
    return hasher.hexdigest()

def get_all_remote_files(access_token, folder_id):
    """Ambil seluruh file di GDrive folder menggunakan REST API v3 dengan pagination."""
    remote_map = {}
    page_token = None
    headers = {"Authorization": f"Bearer {access_token}"}
    print("🔍 Mengambil daftar berkas di Google Drive (dengan pagination)...")
    
    while True:
        url = "https://www.googleapis.com/drive/v3/files"
        params = {
            "q": f"'{folder_id}' in parents and trashed = false",
            "fields": "nextPageToken, files(id, name, webViewLink, md5Checksum)",
            "pageSize": 1000,
            "supportsAllDrives": "true",
            "includeItemsFromAllDrives": "true"
        }
        if page_token:
            params["pageToken"] = page_token
            
        res = requests.get(url, headers=headers, params=params, timeout=30)
        res.raise_for_status()
        data = res.json()
        
        for f in data.get("files", []):
            remote_map[f["name"]] = f
            
        page_token = data.get("nextPageToken")
        if not page_token:
            break
            
    print(f"  ✓ Ditemukan {len(remote_map)} berkas di Google Drive.")
    return remote_map

def upload_file_worker(task):
    """Worker fungsi upload untuk tiap thread."""
    file_path, existing_file, access_token, folder_id = task
    filename = os.path.basename(file_path)
    local_md5 = compute_file_md5(Path(file_path))
    headers = {"Authorization": f"Bearer {access_token}"}
    
    if existing_file:
        remote_md5 = existing_file.get("md5Checksum")
        file_id = existing_file["id"]
        web_link = existing_file.get("webViewLink", f"https://drive.google.com/file/d/{file_id}/view")
        
        if remote_md5 and remote_md5 == local_md5:
            return filename, "SKIP", web_link, None
            
        # Update file in-place (PATCH content)
        update_url = f"https://www.googleapis.com/upload/drive/v3/files/{file_id}?uploadType=media&supportsAllDrives=true"
        with open(file_path, "rb") as f:
            up_res = requests.patch(
                update_url,
                headers={"Authorization": f"Bearer {access_token}", "Content-Type": "application/pdf"},
                data=f,
                timeout=60
            )
        if up_res.status_code == 200:
            return filename, "UPDATE", web_link, None
        else:
            return filename, "FAIL", None, f"Update status {up_res.status_code}: {up_res.text[:200]}"

    # Create new file (Multipart upload)
    create_url = "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&supportsAllDrives=true"
    metadata = {
        "name": filename,
        "parents": [folder_id]
    }
    
    with open(file_path, "rb") as f:
        files_payload = {
            "data": ("metadata", json.dumps(metadata), "application/json; charset=UTF-8"),
            "file": (filename, f, "application/pdf")
        }
        cr_res = requests.post(create_url, headers=headers, files=files_payload, timeout=60)
        
    if cr_res.status_code in (200, 201):
        cr_data = cr_res.json()
        file_id = cr_data["id"]
        web_link = cr_data.get("webViewLink", f"https://drive.google.com/file/d/{file_id}/view")
        
        # Set anyone reader permission
        try:
            perm_url = f"https://www.googleapis.com/drive/v3/files/{file_id}/permissions?supportsAllDrives=true"
            requests.post(perm_url, headers=headers, json={"role": "reader", "type": "anyone"}, timeout=15)
        except Exception:
            pass
            
        return filename, "CREATE", web_link, None
    else:
        return filename, "FAIL", None, f"Create status {cr_res.status_code}: {cr_res.text[:200]}"

def main():
    if len(sys.argv) > 1:
        target_dir = os.path.abspath(sys.argv[1])
    else:
        target_dir = PDF_DIR

    if not os.path.exists(target_dir):
        print(f"❌ Folder tidak ditemukan: {target_dir}")
        return

    pdf_files = sorted([os.path.join(target_dir, f) for f in os.listdir(target_dir) if f.endswith(".pdf")])
    if not pdf_files:
        print(f"⚠️ Tidak ada berkas PDF di {target_dir}")
        return

    print("==================================================================")
    print(f"🚀 GOOGLE DRIVE DELTA UPLOADER — BPS KABUPATEN MEMPAWAH")
    print(f"📁 Folder Lokal : {target_dir} ({len(pdf_files)} PDF)")
    print(f"☁️ Target GDrive : {GDRIVE_FOLDER_ID}")
    print("==================================================================")

    access_token = get_valid_access_token()
    remote_map = get_all_remote_files(access_token, GDRIVE_FOLDER_ID)

    # Siapkan token_user.json untuk kompatibilitas Node.js juga
    try:
        with open(TOKEN_PATH, "r", encoding="utf-8") as f:
            p_token = json.load(f)
        node_token = {
            "access_token": p_token.get("token"),
            "refresh_token": p_token.get("refresh_token"),
            "scope": " ".join(SCOPES),
            "token_type": "Bearer"
        }
        fasih_tok_path = os.path.abspath(os.path.join(BASE_DIR, "..", "fasih-sync-monitoring", "token_user.json"))
        if os.path.exists(os.path.dirname(fasih_tok_path)):
            with open(fasih_tok_path, "w", encoding="utf-8") as f:
                json.dump(node_token, f, indent=2)
    except Exception:
        pass

    link_mapping = {}
    skipped_count = 0
    updated_count = 0
    created_count = 0
    failed_count = 0

    tasks = []
    for fp in pdf_files:
        fn = os.path.basename(fp)
        existing = remote_map.get(fn)
        tasks.append((fp, existing, access_token, GDRIVE_FOLDER_ID))

    print(f"\nMemulai Delta Upload ({len(tasks)} berkas) dengan 12 thread paralel...")

    with ThreadPoolExecutor(max_workers=12) as executor:
        future_to_file = {executor.submit(upload_file_worker, t): t[0] for t in tasks}
        
        for i, future in enumerate(as_completed(future_to_file), 1):
            fp = future_to_file[future]
            fn = os.path.basename(fp)
            code_prefix = fn.split("_")[0] if "_" in fn else fn[:16]
            
            try:
                filename, action, web_link, err = future.result()
                if action == "SKIP":
                    skipped_count += 1
                elif action == "UPDATE":
                    updated_count += 1
                elif action == "CREATE":
                    created_count += 1
                elif action == "FAIL":
                    failed_count += 1
                    print(f"  ❌ [{i}/{len(tasks)}] Gagal {fn}: {err}")
                    
                if web_link and code_prefix:
                    link_mapping[code_prefix] = web_link
                    
            except Exception as e:
                failed_count += 1
                print(f"  ❌ [{i}/{len(tasks)}] Error {fn}: {e}")

            if i % 100 == 0 or i == len(tasks):
                print(f"  Progres: {i}/{len(tasks)} diproses (Skip/Identik: {skipped_count}, Update: {updated_count}, Baru: {created_count}, Gagal: {failed_count})")

    # Simpan mapping link
    for out_dir in [RESULTS_DIR, FASIH_RESULTS_DIR]:
        if os.path.exists(os.path.dirname(out_dir)):
            os.makedirs(out_dir, exist_ok=True)
            map_path = os.path.join(out_dir, "pdf_gdrive_links.json")
            with open(map_path, "w", encoding="utf-8") as f:
                json.dump(link_mapping, f, indent=2)
            print(f"💾 Link mapping disimpan ke: {map_path}")

    print("\n==================================================================")
    print(f"🎉 SUKSES DELTA UPLOAD!")
    print(f"  • Identik (di-Skip) : {skipped_count}")
    print(f"  • Berhasil Update   : {updated_count}")
    print(f"  • Berkas Baru       : {created_count}")
    print(f"  • Gagal             : {failed_count}")
    print(f"  • Total Link Aktif  : {len(link_mapping)}")
    print("==================================================================")

if __name__ == "__main__":
    main()
